packages/compat/src/withNavigationFocus.tsx

```diff
@@ -1,39 +1,44 @@
+/* eslint-disable prettier/prettier */
 import * as React from 'react';
 import hoistNonReactStatics from 'hoist-non-react-statics';
-import { NavigationProp, ParamListBase, useIsFocused } from '@react-navigation/native';
-import type { CompatNavigationProp } from './types';
+import { NavigationProp, useIsFocused } from '@react-navigation/native';
 import useCompatNavigation from './useCompatNavigation';
+import type { NavigationParams, NavigationFocusInjectedProps } from './types';
 
-type InjectedProps<T extends NavigationProp<ParamListBase>> = {
-  isFocused: boolean;
-  navigation: CompatNavigationProp<T>;
-};
+const withNavigation = <
+  C extends React.ComponentType<{}>,
+  NP extends NavigationParams = NavigationParams
+>(
+  Component: C
+) => {
+  type WrappedP = React.ComponentProps<C> & NavigationFocusInjectedProps;
+
+  const componentDisplayName = Component.displayName || Component.name;
+  const WrappedComponent = Component as unknown as React.ComponentType<WrappedP>;
+  const wrappedComponentDisplayName = `withNavigation(${componentDisplayName})`;
 
-export default function withNavigationFocus<
-  T extends NavigationProp<ParamListBase>,
-  P extends InjectedProps<T>,
-  C extends React.ComponentType<P>
->(Comp: C) {
-  const WrappedComponent = ({
-    onRef,
-    ...rest
-  }: Exclude<P, InjectedProps<T>> & {
-    onRef?: C extends React.ComponentClass<any>
-      ? React.Ref<InstanceType<C>>
-      : never;
-  }): React.ReactElement => {
+  const NavigationComponent = React.forwardRef<C, WrappedP>((props, ref) => {
     const isFocused = useIsFocused();
-    const navigation = useCompatNavigation<T>();
+    const navigation = useCompatNavigation<NavigationProp<NP>>();
 
+    return (
     // @ts-expect-error: type checking HOC is hard
-    return <Comp ref={onRef} isFocused={isFocused} navigation={navigation} {...rest} />;
```

Comment:
Spreading `rest` should be before `navigation` because, otherwise, this order means that the parent component can override `navigation`. I don't believe this was intentional.